File: src/main/java/com/leijendary/spring/microservicetemplate/specification/LocalizedSpecification.java
package com.leijendary.spring.microservicetemplate.specification;

import com.leijendary.spring.microservicetemplate.model.LocalizedModel;
import lombok.Builder;
import org.hibernate.query.criteria.internal.OrderImpl;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.lang.NonNull;

import javax.persistence.criteria.*;
import java.util.ArrayList;

import static org.apache.commons.lang3.StringUtils.isBlank;

@Builder
public class LocalizedSpecification<T extends LocalizedModel<?>> implements Specification<T> {

    private final String language;

    @Override
    public Predicate toPredicate(@NonNull final Root<T> root, @NonNull final CriteriaQuery<?> criteriaQuery,
                                 @NonNull final CriteriaBuilder criteriaBuilder) {
        final var predicates = new ArrayList<Predicate>();

        // If there is no language filter, return all based on the reference ID predicate
        if (isBlank(language)) {
            return criteriaQuery.where().getRestriction();
        }

        // ID path for reference ID
        final var idPath = root.<Long>get("id");
        // Join translations for filtering
        final var translationJoin = root.join("translations");

        final var languagePath = translationJoin.<String>get("language");
        // Criteria for exact language
        final var languageEquals = criteriaBuilder.equal(languagePath, language);
        // Parent query ordinal
        final var ordinalPath = translationJoin.<Integer>get("ordinal");

        // Sub query start
        final var type = translationJoin.getJavaType();
        final var ordinalSubQuery = criteriaQuery.subquery(Integer.class);
        final var subQueryRoot = ordinalSubQuery.from(type);
        final var subQueryOrdinalPath = subQueryRoot.<Integer>get("ordinal");
        // Criteria for the lowest ordinal (the default)
        final var subQueryOrdinalMin = criteriaBuilder.min(subQueryOrdinalPath);
        // Filter sub query reference id
        final var subQueryReferenceId = referenceId(subQueryRoot, idPath, criteriaBuilder);
        // Sub query filtering
        ordinalSubQuery.select(subQueryOrdinalMin).where(subQueryReferenceId);

        // Compare the ordinal of the main query vs the sub query
        final var ordinalEqual = criteriaBuilder.equal(ordinalPath, ordinalSubQuery);

        // language OR first ordinal
        final var languageOrFirstOrdinal = criteriaBuilder.or(languageEquals, ordinalEqual);

        predicates.add(languageOrFirstOrdinal);

        // Descending order of ordinal to get the actual language
        // filter first before the default
        final var ordinalDesc = new OrderImpl(ordinalPath).reverse();

        return criteriaQuery
                .where(predicates.toArray(new Predicate[0]))
                .orderBy(ordinalDesc)
                .getRestriction();
    }

    private Predicate referenceId(final Root<?> root, final Path<Long> referenceId,
                                  final CriteriaBuilder criteriaBuilder) {
        // Filter first using the referenceId
        final var referencePath = root.get("reference");
        final var referenceIdPath = referencePath.<Long>get("id");

        return criteriaBuilder.equal(referenceIdPath, referenceId);
    }
}

File: src/main/kotlin/com/leijendary/spring/template/core/specification/LocalizedSpecification.kt
package com.leijendary.spring.template.core.specification

import com.leijendary.spring.template.core.entity.LocalizedEntity
import jakarta.persistence.criteria.*
import org.springframework.data.jpa.domain.Specification

class LocalizedSpecification<T : LocalizedEntity<*>>(val language: String? = null) : Specification<T> {
    override fun toPredicate(
        root: Root<T>,
        criteriaQuery: CriteriaQuery<*>,
        criteriaBuilder: CriteriaBuilder
    ): Predicate {
        val predicates = ArrayList<Predicate>()

        // If there is no language filter, return all based on the reference ID predicate
        if (language.isNullOrBlank()) {
            return criteriaBuilder.and()
        }

        // ID path for reference ID
        val idPath = root.get<Long>("id")
        // Join translations for filtering
        val translationJoin = root.join<Any, Any>("translations")
        val languagePath = translationJoin.get<String>("language")
        // Criteria for exact language
        val languageEquals = criteriaBuilder.equal(languagePath, language)
        // Parent query ordinal
        val ordinalPath = translationJoin.get<Int>("ordinal")

        // Sub query start
        val type = translationJoin.javaType
        val ordinalSubQuery = criteriaQuery.subquery(Int::class.java)
        val subQueryRoot = ordinalSubQuery.from(type)
        val subQueryOrdinalPath = subQueryRoot.get<Int>("ordinal")
        // Criteria for the lowest ordinal (the default)
        val subQueryOrdinalMin = criteriaBuilder.min(subQueryOrdinalPath)
        // Filter sub query reference id
        val subQueryReferenceId = referenceId(subQueryRoot, idPath, criteriaBuilder)

        // Sub query filtering
        ordinalSubQuery.select(subQueryOrdinalMin).where(subQueryReferenceId)

        // Compare the ordinal of the main query vs the sub query
        val ordinalEqual = criteriaBuilder.equal(ordinalPath, ordinalSubQuery)

        // language OR first ordinal
        val languageOrFirstOrdinal = criteriaBuilder.or(languageEquals, ordinalEqual)
        predicates.add(languageOrFirstOrdinal)

        // Descending order of ordinal to get the actual language
        // filter first before the default
        val ordinalDesc = criteriaBuilder.desc(ordinalPath)

        return criteriaQuery
            .where(*predicates.toTypedArray())
            .orderBy(ordinalDesc)
            .restriction
    }

    private fun referenceId(root: Root<*>, referenceId: Path<Long>, criteriaBuilder: CriteriaBuilder): Predicate {
        // Filter first using the referenceId
        val referencePath = root.get<Any>("reference")
        val referenceIdPath = referencePath.get<Long>("id")

        return criteriaBuilder.equal(referenceIdPath, referenceId)
    }
}
